### PR TITLE
Fix lower boundary calculation when pruning range dist table shards

### LINKED ISSUE #5077: Shard pruning might prune incorrect shards of range dist. tables
Reproduced on v8.1.2, not checked older versions.

```sql
CREATE TABLE range_dist_table (
  dist_col BIGINT
);

SELECT create_distributed_table('range_dist_table', 'dist_col', 'range');

select master_create_empty_shard('range_dist_table') as sid \gset
update pg_dist_shard set shardminvalue = 1000 where shardid=:sid;
update pg_dist_shard set shardmaxvalue = 2000 where shardid=:sid;

select master_create_empty_shard('range_dist_table') as sid \gset
update pg_dist_shard set shardminvalue = 3000 where shardid=:sid;
update pg_dist_shard set shardmaxvalue = 4000 where shardid=:sid;

INSERT INTO range_dist_table SELECT i FROM generate_series(1001, 1020) i;
INSERT INTO range_dist_table SELECT i FROM generate_series(3001, 3020) i;
```

```sql
-- evaluates incorrect plan
EXPLAIN SELECT COUNT(*)
FROM range_dist_table
WHERE dist_col >= 2001 AND
      dist_col <= 5000;
┌────────────────────────────────────────────────────────────────┐
│                           QUERY PLAN                           │
├────────────────────────────────────────────────────────────────┤
│ Custom Scan (Citus Adaptive)  (cost=0.00..0.00 rows=0 width=0) │
│   Task Count: 1                                                │
│   Tasks Shown: All                                             │
│   ->  Task                                                     │
│         Node: host=localhost port=10700 dbname=onurctirtir     │
│         ->  Aggregate  (cost=0.01..0.02 rows=1 width=8)        │
│               ->  Result  (cost=0.00..0.01 rows=1 width=0)     │
│                     One-Time Filter: (false AND NULL::boolean) │
└────────────────────────────────────────────────────────────────┘
(8 rows)

-- .. and returns incorrect result
SELECT COUNT(*)
FROM range_dist_table
WHERE dist_col >= 2001 AND
      dist_col <= 5000;
┌───────┐
│ count │
├───────┤
│     0 │
└───────┘
(1 row)
```

```sql
SELECT undistribute_table('range_dist_table');

-- .. where, expected result of the above query is:
SELECT COUNT(*)
FROM range_dist_table
WHERE dist_col >= 2001 AND
      dist_col <= 5000;
┌───────┐
│ count │
├───────┤
│    20 │
└───────┘
(1 row)
```

### PR DESCRIPTION
**Pull request by onurctirtir**
DESCRIPTION: Fixes a bug that causes pruning incorrect shard of a range distributed table

(Will squash & merge the pr)

Fixes #5077.

I would suggest reviewing each commit separately:
* In 36c43fabff14a488aa17312c33c01cff06f851cb, first I made the test that would fail after the bugfix commit more verbose, no functional changes.
* The actual commit that fixes the bug is cebba0596f62c52cb65513acca5117b453abc868.
* 45738de610468497ba014b13936af39fdafcef93 & c3c6c23f6058d8d3a70f30f37dc906b4acc9fe4b are just to clarify what we are doing, no functional changes.

This happens only when we have a "<" or "<=" filter on distribution
column of a range distributed table and that filter falls in between
two shards.

When the filter falls in between two shards:

* If the filter is ">" or ">=", then UpperShardBoundary was
    returning "upperBoundIndex - 1", where upperBoundIndex is
    exclusive shard index used during binary seach.
    This is expected since upperBoundIndex is an exclusive
    index.

* If the filter is "<" or "<=", then LowerShardBoundary was
    returning "lowerBoundIndex + 1", where lowerBoundIndex is
    inclusive shard index used during binary seach.
    On the other hand, since lowerBoundIndex is an inclusive
    index, we should just return lowerBoundIndex instead of
    doing "+ 1". Before this commit, we were missing leftmost
    shard in such queries.

### CODE REVIEW

File: src/backend/distributed/planner/shard_pruning.c [hunk 1]
```diff
@@ -1614,20 +1630,30 @@ LowerShardBoundary(Datum partitionColumnValue, ShardInterval **shardIntervalCach
...
-		/* all intervals are bigger, thus return 0 */
-		return 0;
-	}
-	else if (lowerBoundIndex == shardCount)
+	if (lowerBoundIndex == shardCount)
 	{
-		/* partition value is bigger than all partition values */
+		/*
+		 * Since lowerBoundIndex is an inclusive index, INVALID_SHARD_INDEX
+		 * means all the shards have smaller values than partitionColumnValue,
```
Comment (JelteF) [inline src/backend/distributed/planner/shard_pruning.c:1642]:
```suggestion
		 * Since lowerBoundIndex is an inclusive index, if it equal to shardCount then that
		 * means all the shards have smaller values than partitionColumnValue,
```

File: src/backend/distributed/planner/shard_pruning.c [hunk 2]
```diff
@@ -1571,6 +1571,22 @@ LowerShardBoundary(Datum partitionColumnValue, ShardInterval **shardIntervalCach
...
+	 *    partitionColumnValue >= shard[x].min, and
+	 *    partitionColumnValue < shard[x].max (or partitionColumnValue <= shard[x].max).
+	 * 2) partitionColumnValue < shard[x].min for all the shards
+	 * 3) partitionColumnValue > shard[x].max for all the shards
+	 * 4) partitionColumnValue falls in between two shards, such that:
+	 *    partitionColumnValue > shard[x].max and
+	 *    partitionColumnValue < shard[x+1].min
+	 *
+	 * For 1), we find that shard in below loop and return the index of it.
+	 * For the others, see the end of this function.
```
Comment (JelteF) [inline src/backend/distributed/planner/shard_pruning.c:1588]:
```suggestion
	 * For 1), we find that shard in below loop using binary search and return the index of it.
	 * For the others, see the end of this function.
```

File: src/backend/distributed/planner/shard_pruning.c [hunk 3]
```diff
@@ -1647,6 +1673,23 @@ UpperShardBoundary(Datum partitionColumnValue, ShardInterval **shardIntervalCach
...
+	 *    partitionColumnValue <= shard[x].max, and
+	 *    partitionColumnValue > shard[x].min (or partitionColumnValue >= shard[x].min).
+	 * 2) partitionColumnValue > shard[x].max for all the shards
+	 * 3) partitionColumnValue < shard[x].min for all the shards
+	 * 4) partitionColumnValue falls in between two shards, such that:
+	 *    partitionColumnValue > shard[x].max and
+	 *    partitionColumnValue < shard[x+1].min
+	 *
+	 * For 1), we find that shard in below loop and return the index of it.
+	 * For the others, see the end of this function.
```
Comment (JelteF) [inline src/backend/distributed/planner/shard_pruning.c:1690]:
```suggestion
	 * For 1), we find that shard in below loop using binary search and return the index of it.
	 * For the others, see the end of this function.
```

**Review by JelteF — APPROVED** [review]

**Review by JelteF — APPROVED** [review]

File: src/test/regress/expected/multi_prune_shard_list.out [hunk 4]
```diff
@@ -543,6 +543,130 @@ SELECT * FROM numeric_test WHERE id = 21.1::numeric;
...
+ t
+(1 row)
+
+SELECT SUM(dist_col)=3800+6500 FROM range_dist_table_1 WHERE dist_col >= 2500;
+ ?column?
+---------------------------------------------------------------------
+ t
+(1 row)
+
+SELECT SUM(dist_col)=3800+6500 FROM range_dist_table_1 WHERE dist_col > 2000;
```
Comment (onderkalaci) [inline src/test/regress/expected/multi_prune_shard_list.out:576]: can you add three more tests with > 1001, >=1001, >= 1000 and > 1000

File: src/test/regress/expected/multi_prune_shard_list.out [hunk 5]
```diff
@@ -543,6 +543,130 @@ SELECT * FROM numeric_test WHERE id = 21.1::numeric;
...
+CREATE TYPE comp_type_range AS RANGE (
+    subtype = comp_type);
+CREATE TABLE range_dist_table_2 (dist_col comp_type);
+SELECT create_distributed_table('range_dist_table_2', 'dist_col', 'range');
+ create_distributed_table
+---------------------------------------------------------------------
+
+(1 row)
+
+CALL public.create_range_partitioned_shards(
```
Comment (onderkalaci) [inline src/test/regress/expected/multi_prune_shard_list.out:622]: minor: These values are very large integers and hard to follow. Can you make them smaller values so that the test becomes more readable? To be fair, I couldn't validate the results by looking this test

**Review by onderkalaci — APPROVED** [review]

**Comment by onurctirtir** [comment]
Already backported to 10.1 too, so removing the label
